pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -1172,40 +878,59 @@ func (pl *dynamicResources) Filter(ctx context.Context, cs *framework.CycleState
+			if !nodeSelector.Match(node) {
+				return statusUnschedulable(logger, "excluded by device class node filter", "pod", klog.KObj(pod), "node", klog.KObj(node), "deviceclass", klog.KRef("", className))
 			}
 		}
-		// Can the builtin controller tell us whether the node is suitable?
-		if state.informationsForClaim[index].structuredParameters {
-			suitable, err := state.informationsForClaim[index].controller.nodeIsSuitable(ctx, node.Name, state.resources)
-			if err != nil {
-				// An error indicates that something wasn't configured correctly, for example
-				// writing a CEL expression which doesn't handle a map lookup error. Normally
-				// this should never fail. We could return an error here, but then the pod
-				// would get retried. Instead we ignore the node.
-				return statusUnschedulable(logger, fmt.Sprintf("checking structured parameters failed: %v", err), "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaim", klog.KObj(claim))
-			}
-			if !suitable {
-				return statusUnschedulable(logger, "resourceclaim cannot be allocated for the node (unsuitable)", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaim", klog.KObj(claim))
-			}
-		} else {
-			if status := state.informationsForClaim[index].status; status != nil {
-				for _, unsuitableNode := range status.UnsuitableNodes {
-					if node.Name == unsuitableNode {
-						return statusUnschedulable(logger, "resourceclaim cannot be allocated for the node (unsuitable)", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaim", klog.KObj(claim), "unsuitablenodes", status.UnsuitableNodes)
-					}
+
+		// Use information from control plane controller?
+		if status := state.informationsForClaim[index].status; status != nil {
+			for _, unsuitableNode := range status.UnsuitableNodes {
+				if node.Name == unsuitableNode {
+					return statusUnschedulable(logger, "resourceclaim cannot be allocated for the node (unsuitable)", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaim", klog.KObj(claim), "unsuitablenodes", status.UnsuitableNodes)
 				}
 			}
 		}
 	}
 
+	// Use allocator to check the node and cache the result in case that the node is picked.
+	var allocations []*resourceapi.AllocationResult
+	if state.allocator != nil {
+		allocCtx := ctx
+		if loggerV := logger.V(5); loggerV.Enabled() {
+			allocCtx = klog.NewContext(allocCtx, klog.LoggerWithValues(logger, "node", klog.KObj(node)))
+		}
+
+		a, err := state.allocator.Allocate(allocCtx, node)
+		if err != nil {
+			// This should only fail if there is something wrong with the claim or class.
+			// Mark the pod as unschedulable until they get updated to fix the problem.
+			//
+			// Also record the error in the state to ensure that PreScore sees it and
+			// can prevent scheduling from continuing.
+			state.mutex.Lock()
+			defer state.mutex.Unlock()
+			state.filterErr = fmt.Errorf("filter node %s: %w", node.Name, err)
```

Comment:
Is this error specific w/ particular Node? if so, why we put it as a single file in stateData; if not, e.g., if it's just CEL syntax error, maybe we catch potential error earlier (like in PreFilter) so that it doesn't need to be repeatedly checked and failed in Filter()?

Reply:
See https://github.com/kubernetes/kubernetes/pull/125488/files#r1686027537